code/lib/preview-api/src/modules/store/csf/beforeAll.ts

```diff
@@ -0,0 +1,21 @@
+type Awaitable<T> = T | Promise<T>;
+
+type CleanupFn = () => Awaitable<void>;
+
+export type BeforeAllHook = () => Awaitable<void | CleanupFn>;
+
+// Execute all the hooks in sequence, and return a function that will execute cleanups in reverse order
+export const composeBeforeHooks = (hooks: BeforeAllHook[]): BeforeAllHook => {
```

Comment:
`composeBeforeAllHooks`?

Reply:
Yeah I wasn't sure whether we'd want to reuse this for `beforeEach`. WDYT?

Reply:
I kind of like the idea, but I'm changing beforeEach in my PR, to short circuit when the story is aborted, so would be kind of less straight-forward to make this generic:

```ts
  const applyBeforeEach = async (context: StoryContext<TRenderer>): Promise<CleanupCallback[]> => {
    const cleanupCallbacks = new Array<() => unknown>();
    for (const beforeEach of [
      ...normalizeArrays(projectAnnotations.beforeEach),
      ...normalizeArrays(componentAnnotations.beforeEach),
      ...normalizeArrays(storyAnnotations.beforeEach),
    ]) {
      if (context.abortSignal.aborted) return cleanupCallbacks;
      const cleanup = await beforeEach(context);
      if (cleanup) cleanupCallbacks.push(cleanup);
    }
    return cleanupCallbacks;
  };
  ```
I would say, leave this to beforeAll for now.